### PR TITLE
Improve "Dev contracts"

### PR DESCRIPTION
**Pull request by EugenWay**
Updated sections:

-
-
-

### CODE REVIEW

File: docs/developing-contracts/gear-program.md [hunk 1]
```diff
@@ -1,29 +1,52 @@
...
-sidebar_position: 2
+sidebar_label: Executable functions
+sidebar_position: 1
 ---
 
-# Gear program state
+# Executable functions
 
-The program is the main unit of the Gear component. Program code is stored as an immutable WASM blob. Each program has a fixed amount of memory which persists between messagehandling (so-called static area).
+The program is the main unit of the Gear component. Each program is immutable [WASM](/docs/gear/technology/WASM) blob and has a fixed amount of memory which persists between messagehandling (so-called static area).
```
[the remark this answers is not in the document]
Comment (breathx) [reply docs/developing-contracts/gear-program.md:8]: And if we tell about static area, would be good to also mention dynamic one

File: docs/developing-contracts/gstd.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+sidebar_position: 2
+---
+
+# Gear library
+
+Gear library `gstd` provides all the necessary and sufficient functions and methods for develop smart-contracts.
+
+## Message handing
+
+Gear allows users and programs to interact with other users and programs via messages. Messages can contain a `payload` that will be able to be processed during message executione. Interaction with messages is possible thanks to the module `msg`
```
Comment (breathx) [inline docs/developing-contracts/gstd.md:12]:
```suggestion
Gear allows users and programs to interact with other users and programs via messages. Messages can contain a `payload` that will be able to be processed during message execution. Interaction with messages is possible thanks to the module `msg`
```

File: docs/developing-contracts/gstd.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    // ...
+    let id = msg::source();
+    let message_string: String = "Hello there"
+    msg::send(id, message_string, 0).expect("Unable to send message");
+}
+```
+
+You can see more cases of using `msg` module in our [docs](https://docs.gear.rs/gstd/msg/index.html)
+
+## Sys call's
```
Comment (breathx) [inline docs/developing-contracts/gstd.md:56]:
```suggestion
## Execution control
```

File: docs/developing-contracts/state.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+Persistent data of the Gear contact is stored in the same way as in a classic program and does not require initialization of the external storage.
+
+```rust
+// describe state structure
+pub struct State {
+    votes_received: BTreeMap<String, i32>,
+}
+
+// initialize state itsalfe
+static mut STATE: State = State::new();
```
Comment (breathx) [inline docs/developing-contracts/state.md:17]: This code is invalid.
1. State has no `new` method explicitly.
2. `new` fn while appear should be `const fn` -> BTreeMap const constructor is unstable and provided by nightly feature

File: docs/developing-contracts/state.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+```
+
+If you program in Rust or other object-oriented languages, then you should be familiar with most types. However `ActorId` type shoud be new in Gear contract development.
+
+:::info
+`ActorId` is a special type 
+:::
```
Comment (breathx) [inline docs/developing-contracts/state.md:24]: Strange remark for me

Comment (breathx) [reply docs/developing-contracts/state.md:24]: Probably, it needs some more details to be more informative
E.g. it's 32 bytes array defining id on a blockchain

File: docs/developing-contracts/state.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+## State functions
+
+To display the contact status information, by analogy with the "view" Gear functions, `meta_state()` is used, which allows you to instantly read the contact status (for example, balance) without any costs.
+
+```rust
+// The function that returns a part of memory with a state
+#[no_mangle]
+pub unsafe extern "C" fn meta_state() -> *mut [i32; 2] {
+
```
Comment (breathx) [inline docs/developing-contracts/state.md:34]: Please, provide same dummy example. It needs to specify usage of `to_leak_ptr`

File: docs/developing-contracts/codec.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+```toml
+[dependencies]
+
+// ...
+codec = { package = "parity-scale-codec", version = "3.1.2", default-features = false, features = ["derive", "full"] }
+```
+
+```rust
+#[derive(Encode, Decode)]
```
Comment (breathx) [inline docs/developing-contracts/codec.md:20]:
```suggestion
```rust
use codec::{Decode, Encode};

#[derive(Decode, Encode)]
```

File: docs/developing-contracts/codec.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+The definitions provide third party tools (e.g. a UI client) with information about how they are able to decode types agnostic of language.
+
+```toml
+[dependencies]
+
+// ...
+scale-info = { version = "2.2.0", default-features = false, features = ["derive"] }
+```
+
+[Learn more about scale-info](https://github.com/paritytech/scale-info)
```
Comment (breathx) [inline docs/developing-contracts/codec.md:42]: duplicate info this section

File: docs/developing-contracts/codec.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+codec = { package = "parity-scale-codec", version = "3.1.2", default-features = false, features = ["derive", "full"] }
+```
+
+```rust
+#[derive(Encode, Decode)]
+enum MyType {
+    MyStruct { field: ... },
+    ...
+}
+```
```
Comment (breathx) [inline docs/developing-contracts/codec.md:25]: Seems good to mention that we use Encode and Decode trait bound in some gstd methods for soft wrapping (it's difference between e.g. `send` and `send_bytes`)

File: docs/developing-contracts/interactions-between-programs.md [hunk 10]
```diff
@@ -1,197 +1,54 @@
...
-gstd::metadata! {
-    title: "MyFungibleToken",
-    init:
-        input: InitConfig,
-    handle:
-        input: Action,
-        output: Event,
-}
-```
+# Cross-program communication
```
Comment (breathx) [inline docs/developing-contracts/interactions-between-programs.md:6]: Cross program invocation is not about async at all. It suits under telling about actor model. I'd provide separate section for async story in gear protocol

File: docs/developing-contracts/create.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+```rust
+use gstd::{prog::ProgramGenerator, CodeHash};
+
+#[no_mangle]
+unsafe extern "C" fn handle() {
+    let submitted_code: CodeHash =
+        hex_literal::hex!("abf3746e72a6e8740bd9e12b879fbdd59e052cb390f116454e9116c22021ae4a")
+            .into();
+
+    ProgramGenerator::create_program_with_gas(submitted_code, b"payload", 10_000_000_000, 0)
```
Comment (breathx) [inline docs/developing-contracts/create.md:27]: It returns program id for future usage. What about to send it message right after initialisation?

File: docs/developing-contracts/codec.md (hunk 8 above)
Comment (EugenWay) [reply docs/developing-contracts/codec.md:67]: Added metadata example

File: docs/developing-contracts/codec.md (hunk 9 above)
Comment (EugenWay) [reply docs/developing-contracts/codec.md:27]: Left a note about the difference

File: docs/developing-contracts/interactions-between-programs.md (hunk 10 above)
Comment (EugenWay) [reply docs/developing-contracts/interactions-between-programs.md:6]: Any suggestions?

**Review by AndrePanin — APPROVED** [review]

File: docs/developing-contracts/interactions-between-programs.md (hunk 10 above)
Comment (EugenWay) [reply docs/developing-contracts/interactions-between-programs.md]: Done

**Pull request merged**